
250318 Test
updated version with no primer dimer, on same 231 GEMs

```bash
#first need to make the output dir and the log directory for bcl-convert
outdir="/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231"
mkdir -p ${outdir}
mkdir -p ${outdir}/logs

cd /home/rmulqueen/projects/kismet/ #move to project directory
git clone https://github.com/mulqueenr/scmet_nf_processing ./tools/scmet_nf_processing #pull github repo
```

# RM03 10N

```bash
#index is scalebio homebrew set A9
#TATAGCGCGG #trimming first 2 index cycles (i think that's right?)
#reverse complement then trim off last 2 bases from index sequences
#CCGCGCTATA #CCGCGCTA

nextflow ./tools/scmet_nf_processing/nextflow_running/kismet_processing.groovy \
--flowcellDir /home/rmulqueen/projects/kismet/seq/250318_VH01788_96_AAGHCK2M5 \
--outname RM03_10N \
--outdir ${outdir} \
--i7_idx CCGCGCTA \
--sequencing_cycles "Y50;I8;U16;Y50" \
--max_cpus 300 \
-with-report \
-resume

```

# RM04 10H
```bash

#index is scalebio homebrew set A10
#CGCAAGTATT
#AATACTTGCG #AATACTTG revcomp, trim 2
nextflow ./tools/scmet_nf_processing/nextflow_running/kismet_processing.groovy \
--flowcellDir /home/rmulqueen/projects/kismet/seq/250318_VH01788_96_AAGHCK2M5 \
--outname RM04_10H \
--outdir ${outdir} \
--i7_idx AATACTTG \
--sequencing_cycles "Y50;I8;U16;Y50" \
--max_cpus 300 \
-with-report 
```

# Regular 10x cellranger-atac for RM02 (SI-NA-B10)

```bash
cd ${outdir}

ref="/home/rmulqueen/ref/refdata-cellranger-arc-GRCh38-2020-A-2.0.0"

echo """Lane,Sample,Index
*,RM02_10x,SI-NA-B10""" > RM02_10x.samplesheet.csv

~/tools/cellranger-atac-2.1.0/cellranger-atac mkfastq \
--id=RM02_10x \
--run=/home/rmulqueen/projects/kismet/seq/250318_VH01788_96_AAGHCK2M5 \
--csv=RM02_10x.samplesheet.csv

~/tools/cellranger-atac-2.1.0/cellranger-atac count \
--id=RM02_10x_count \
--reference=${ref} \
--fastqs=/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231/RM02_10x/outs/fastq_path/AAGHCK2M5/RM02_10x \
--sample=RM02_10x \
--localcores=300 \
--localmem=1000

#make splitting barcode list from whitelist
outdir="/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231/cellranger_dna/sc_bams"
mkdir -p $outdir
indir="/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231/cellranger_dna"
dna_barcodes="${indir}/RM02_10x_count/outs/filtered_peak_bc_matrix/barcodes.tsv"
bam="${indir}/RM02_10x_count/outs/possorted_bam.bam"

awk -F, 'OFS="\t" {print $1,$1}' ${dna_barcodes} > ${outdir}/cell_id.tsv

cd $outdir
#split to chunks of 500 cells for i/o purposes
split -l 500 --numeric-suffixes ${outdir}/cell_id.tsv ${outdir}/cell_id.split.

#run cell splitting for each 500 chunk
for i in cell_id.split* ; do
sinto filterbarcodes --bam ${bam} --cells $i -p 100 --barcodetag "CB" --outdir ./sc_dna_bam ;
done


#generate library complexity based on 10% downsample rates
#count unique chr:start sites
function proj_complexity() {
cellid="RM02_10x_${1::-6}"
for i in $(seq 0.1 0.1 1.0); do
uniq_count=$(samtools view -F 3332 -s $i $1 \
| awk 'OFS="\t"{print $3,$4}' \
| sort \
| uniq -c \
| wc -l)
total_count=$(samtools view -F 3332 -s $i $1 | wc -l)
echo "${cellid},${i},${total_count},${uniq_count}"; done > ${cellid}.projected_metrics.txt
}

export -f proj_complexity
parallel -j 100 proj_complexity ::: $(ls *-1.bam)

#excluding reads that meet any below conditions:
#read unmapped (0x4)
#not primary alignment (0x100)
#read is PCR or optical duplicate (0x400)
#supplementary alignment (0x800)


```