
250318 Test
updated version with no primer dimer, on same 231 GEMs

```bash
#first need to make the output dir and the log directory for bcl-convert
outdir="/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231"
mkdir -p ${outdir}
mkdir -p ${outdir}/logs

cd /home/rmulqueen/projects/kismet/ #move to project directory
git clone https://github.com/mulqueenr/scmet_nf_processing ./tools/scmet_nf_processing #pull github repo
```

# RM03 10N

```bash
#index is scalebio homebrew set A9
#TATAGCGCGG #trimming first 2 index cycles (i think that's right?)
#reverse complement then trim off last 2 bases from index sequences
#CCGCGCTATA #CCGCGCTA

nextflow ./tools/scmet_nf_processing/nextflow_running/kismet_processing.groovy \
--flowcellDir /home/rmulqueen/projects/kismet/seq/250318_VH01788_96_AAGHCK2M5 \
--outname RM03_10N \
--outdir ${outdir} \
--i7_idx CCGCGCTA \
--sequencing_cycles "Y50;I8;U16;Y50" \
--max_cpus 300 \
-with-report \
-resume

```

# RM04 10H
```bash

#index is scalebio homebrew set A10
#CGCAAGTATT
#AATACTTGCG #AATACTTG revcomp, trim 2
nextflow ./tools/scmet_nf_processing/nextflow_running/kismet_processing.groovy \
--flowcellDir /home/rmulqueen/projects/kismet/seq/250318_VH01788_96_AAGHCK2M5 \
--outname RM04_10H \
--outdir ${outdir} \
--i7_idx AATACTTG \
--sequencing_cycles "Y50;I8;U16;Y50" \
--max_cpus 300 \
-with-report 
```

# Regular 10x cellranger-atac for RM02 (SI-NA-B10)

```bash
cd ${outdir}

ref="/home/rmulqueen/ref/refdata-cellranger-arc-GRCh38-2020-A-2.0.0"

echo """Lane,Sample,Index
*,RM02_10x,SI-NA-B10""" > RM02_10x.samplesheet.csv

~/tools/cellranger-atac-2.1.0/cellranger-atac mkfastq \
--id=RM02_10x \
--run=/home/rmulqueen/projects/kismet/seq/250318_VH01788_96_AAGHCK2M5 \
--csv=RM02_10x.samplesheet.csv

~/tools/cellranger-atac-2.1.0/cellranger-atac count \
--id=RM02_10x_count \
--reference=${ref} \
--fastqs=/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231/RM02_10x/outs/fastq_path/AAGHCK2M5/RM02_10x \
--sample=RM02_10x \
--localcores=300 \
--localmem=1000

#make splitting barcode list from whitelist
outdir="/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231/cellranger_dna/sc_bams"
mkdir -p $outdir
indir="/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231/cellranger_dna"
dna_barcodes="${indir}/RM02_10x_count/outs/filtered_peak_bc_matrix/barcodes.tsv"
bam="${indir}/RM02_10x_count/outs/possorted_bam.bam"

awk -F, 'OFS="\t" {print $1,$1}' ${dna_barcodes} > ${outdir}/cell_id.tsv

cd $outdir
#split to chunks of 500 cells for i/o purposes
split -l 500 --numeric-suffixes ${outdir}/cell_id.tsv ${outdir}/cell_id.split.

#run cell splitting for each 500 chunk
for i in cell_id.split* ; do
sinto filterbarcodes --bam ${bam} --cells $i -p 100 --barcodetag "CB" --outdir ./sc_dna_bam ;
done


#generate library complexity based on 10% downsample rates
#count unique chr:start sites
function proj_complexity() {
cellid="${1::-4}"
for i in $(seq 0.1 0.1 1.0); do
uniq_count=$(samtools view -F 3332 -s $i $1 \
| awk 'OFS="\t"{print $3,$4}' \
| sort \
| uniq -c \
| wc -l)
total_count=$(samtools view -F 3332 -s $i $1 | wc -l)
echo "${cellid},${i},${total_count},${uniq_count}"; done > ${cellid}.projected_metrics.txt
}

export -f proj_complexity
parallel -j 100 proj_complexity ::: $(ls *.bam)

cat *projected_metrics.txt > ../proj_read_counts.txt
#excluding reads that meet any below conditions:
#read unmapped (0x4)
#not primary alignment (0x100)
#read is PCR or optical duplicate (0x400)
#supplementary alignment (0x800)


```


```R
library(ggplot2)
library(patchwork)
library(drc)
library(parallel)
setwd("/volumes/USR2/Ryan")

scale<-read.table("scalebio_proj_read_counts.txt",sep=",")
colnames(scale)<-c("cellid","downsamp_perc","total_reads","uniq_reads")
scale$sample<-unlist(lapply(strsplit(scale$cellid,"[.]"),"[",1))
scale$method<-"scale"
scale<-scale[!(scale$cellid=="MDA231__sc"),]
kismet<-read.table("kismet_proj_read_counts.txt",sep=",")
colnames(kismet)<-c("cellid","downsamp_perc","total_reads","uniq_reads")
kismet$sample<-"MDA-MB-231"
kismet$method<-"kismet"

compl<-rbind(scale,kismet)

michaelis_menten_fit<-function(x){
    mm<-compl[compl$cellid==x,]
    colnames(mm)<-c("cellid","downsample_perc","S","v","sample","method")
    model.drm <- drm(v ~ S, data = mm, fct = MM.2())
    km_uniq <- data.frame(S = coef(model.drm)[2])
    km_uniq$v <- predict(model.drm, newdata = km_uniq)
    vmax<-as.numeric(coef(model.drm)[1])
    km<-as.numeric(coef(model.drm)[2])
    current_total_reads<-as.numeric(mm[mm$downsample_perc==1.0,]$S)
    current_uniq_reads<-as.numeric(mm[mm$downsample_perc==1.0,]$v)
    method<-mm[mm$downsample_perc==1.0,]$method
    return(c(x,vmax,km,km_uniq$v,current_total_reads,current_uniq_reads,method))
}

projdat<-as.data.frame(do.call("rbind",mclapply(mc.cores=100,unique(compl$cellid),michaelis_menten_fit)))


colnames(projdat)<-c("sample",
"projected_total_fragments",
"projected_optimal_seq_effort",
"projected_reads_at_optimal_effort",
"current_total_reads",
"current_uniq_reads",
"method")

#filter cells with less than 10000 current reads
projdat$current_reads<-as.numeric(projdat$current_reads)

#project complexity for cells passing QC
dat<-read.table("proj_read_counts.txt",header=F,sep=",")
colnames(dat)<-c("sample","downsample_perc","total_frag","uniq_frag")
dat$method<-unlist(lapply(strsplit(dat$sample,"_[ATCG][ATCG]"),"[",1))

#these are already sequence exhausted, so don't need projections (also DRC isnt installed)
dat<-dat[dat$downsample_perc==0.50,]
dat$perc_uniq<-dat$uniq_frag/dat$total_frag

plt1<-ggplot(dat,aes(x=method,y=log10(as.numeric(uniq_frag)),alpha=0.5,color=method))+
geom_jitter()+
theme_minimal()+
ylim(c(0,7))+
ggtitle(paste0("Current uniq perc effort:\n","Mean: ",round(mean(dat$perc_uniq*100))))
ggsave(plt1,file="uniq_reads_per_method.pdf")

library(dplyr)
dat %>% group_by(method) %>% summarize(count=n(),reads=mean(uniq_frag),perc_uniq=mean(perc_uniq))

```