
250424 Test
updated version trying to increase efficiency, barnyard

```bash
#first need to make the output dir and the log directory for bcl-convert
outdir="/data/rmulqueen/projects/kismet/data/250905_kismetv51_optimized"
mkdir -p ${outdir}
mkdir -p ${outdir}/logs

cd /data/rmulqueen/projects/kismet/ #move to project directory
git clone https://github.com/mulqueenr/scmet_nf_processing ./tools/scmet_nf_processing #pull github repo
```

# Optmized Kismet v5.1

```bash
#index is scalebio homebrew set A9
#TATAGCGCGG /Volumes/data/rmulqueen/projects/kismet/design/10x_met_design_250417.xlsx
#do rev comp, trim first 2 for i7
#run on hg38 (MDA-MB-231)

cd /data/rmulqueen/projects/kismet/ #move to project directory

nextflow /data/rmulqueen/projects/kismet/tools/scmet_nf_processing/nextflow_running/kismet_processing.groovy \
--flowcellDir /data/rmulqueen/projects/kismet/seq/250904_VH01788_122_AAHFTCNM5 \
--outname kismet_optimized \
--outdir $outdir \
--ref_index /home/rmulqueen/ref/hg38_bsbolt \
--i7_idx CCGCGCTA \
--sequencing_cycles "Y50;I8;U16;Y50" \
--max_cpus 100 
```



```bash

#generate library complexity based on 10% downsample rates
#count unique chr:start sites
function barnyard_count() {
cellid="${1::-9}"
samtools view -F 3332 $1 | awk '{if(substr($3,0,6) == "GRCm39"){print "mouse"} else {print "human"}}' \
| sort | uniq -c \
| awk -v in_bam=$cellid 'OFS="\t" {print $0,in_bam}'
}

export -f barnyard_count
parallel -j 100 barnyard_count ::: $(ls *.bam) > barnyard.readcount.tsv

```
```R
library(ggplot2)
library(reshape2)
setwd("/data/rmulqueen/projects/kismet/data/250424_kismetv5_barnyard")

dat<-read.table("barnyard.readcount.tsv")
colnames(dat)<-c("count","species","cell")
dat_cast<-dcast(dat,cell~species,value.var='count',fill=0)
dat_cast$method<-unlist(lapply(strsplit(dat_cast$cell,"_[ATCG][ATCG]"),"[",1))

plt<-ggplot(dat_cast,aes(x=human,y=mouse,color=method))+geom_point()+theme_minimal()
ggsave(plt,file="kismet_barnyard.method.pdf")

dat_cast<-dat_cast[dat_cast$mouse+dat_cast$human>500,]
dat_cast$perc<-dat_cast$human/(dat_cast$mouse+dat_cast$human)
dat_cast$collision<-ifelse(dat_cast$perc<0.90 & dat_cast$perc>0.10,"collision","singleton")
plt<-ggplot(dat_cast,aes(x=human,y=mouse,color=collision))+geom_point()+theme_minimal()
ggsave(plt,file="kismet_barnyard.collision.pdf")

plt<-ggplot(dat_cast,aes(x=log10(human),y=log10(mouse),color=collision))+geom_point()+theme_minimal()
ggsave(plt,file="kismet_barnyard.collision.log.pdf")

collision_rates<-data.frame(rbind(table(dat_cast$method,dat_cast$collision)))
(collision_rates$collision/collision_rates$singleton)*2
#[1] 0.0000000 0.2913386 0.2231405 0.4199475
```
Project complexity
```R
library(ggplot2)
library(patchwork)
#library(drc)
library(parallel)

#project complexity for cells passing QC
dat<-do.call("rbind",lapply(list.files(path="./reports/projected_size/",full.names=T),function(x) read.table(x,header=F,sep=",")))
colnames(dat)<-c("sample","downsample_perc","total_frag","uniq_frag")
dat$method<-unlist(lapply(strsplit(dat$sample,"_[ATCG][ATCG]"),"[",1))

#these are already sequence exhausted, so don't need projections (also DRC isnt installed)
dat<-dat[dat$downsample_perc==1.0,]
dat$perc_uniq<-dat$uniq_frag/dat$total_frag

plt1<-ggplot(dat,aes(x=method,y=log10(as.numeric(uniq_frag)),alpha=0.5,color=method))+
geom_jitter()+
theme_minimal()+
ylim(c(0,7))+
ggtitle(paste0("Current uniq perc effort:\n","Mean: ",round(mean(dat$perc_uniq*100))))
ggsave(plt1,file="uniq_reads_per_method.pdf")

library(dplyr)
dat %>% group_by(method) %>% summarize(count=n(),reads=mean(uniq_frag),perc_uniq=mean(perc_uniq))
```
# Regular 10x cellranger-atac for 1_5 (SI-NA-G11)
Run in amethyst.sif
```bash
singularity shell \
--bind /data/rmulqueen/projects/kismet \
--bind /home/rmulqueen/ref/ \
--bind /home/rmulqueen:/var/log/bcl-convert \
/home/rmulqueen/singularity/amethyst.sif

source /container_src/container_bashrc

outdir="/data/rmulqueen/projects/kismet/data/250424_kismetv5_barnyard"
cd ${outdir}

ref="/home/rmulqueen/ref/refdata-cellranger-arc-GRCh38-and-GRCm39-2024-A/"

echo """[Header]
FileFormatVersion,2
[BCLConvert_Settings]
CreateFastqForIndexReads,0
OverrideCycles,Y50;I8;U16;Y50
[BCLConvert_Data]
Sample_ID,index
1_5_BY_10x,AGGGCGTT
1_5_BY_10x,CTATACGC
1_5_BY_10x,GCTCGTCA
1_5_BY_10x,TACATAAG""" > 1_5_BY_10x.samplesheet.csv

bcl-convert \
--bcl-input-directory /data/rmulqueen/projects/kismet/seq/250423_VH01788_101_AAGWNHVM5 \
--output-directory /data/rmulqueen/projects/kismet/data/250424_kismetv5_barnyard/1_5_BY_10x \
--sample-sheet 1_5_BY_10x.samplesheet.csv --force

~/tools/cellranger-atac-2.1.0/cellranger-atac count \
--id=1_5_BY_10x \
--reference=${ref} \
--fastqs=/data/rmulqueen/projects/kismet/data/250424_kismetv5_barnyard \
--sample=1_5_BY_10x \
--project=1_5_BY_10x \
--localcores=300 \
--localmem=1000

#make splitting barcode list from whitelist
outdir="/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231/cellranger_dna/sc_bams"
mkdir -p $outdir
indir="/home/rmulqueen/projects/kismet/data/250318_kismetv4_2_231/cellranger_dna"
dna_barcodes="${indir}/RM02_10x_count/outs/filtered_peak_bc_matrix/barcodes.tsv"
bam="${indir}/RM02_10x_count/outs/possorted_bam.bam"

awk -F, 'OFS="\t" {print $1,$1}' ${dna_barcodes} > ${outdir}/cell_id.tsv

cd $outdir
#split to chunks of 500 cells for i/o purposes
split -l 500 --numeric-suffixes ${outdir}/cell_id.tsv ${outdir}/cell_id.split.

#run cell splitting for each 500 chunk
for i in cell_id.split* ; do
sinto filterbarcodes --bam ${bam} --cells $i -p 100 --barcodetag "CB" --outdir ./sc_dna_bam ;
done


#generate library complexity based on 10% downsample rates
#count unique chr:start sites
function proj_complexity() {
cellid="RM02_10x_${1::-6}"
for i in $(seq 0.1 0.1 1.0); do
uniq_count=$(samtools view -F 3332 -s $i $1 \
| awk 'OFS="\t"{print $3,$4}' \
| sort \
| uniq -c \
| wc -l)
total_count=$(samtools view -F 3332 -s $i $1 | wc -l)
echo "${cellid},${i},${total_count},${uniq_count}"; done > ${cellid}.projected_metrics.txt
}

export -f proj_complexity
parallel -j 100 proj_complexity ::: $(ls *-1.bam)

#excluding reads that meet any below conditions:
#read unmapped (0x4)
#not primary alignment (0x100)
#read is PCR or optical duplicate (0x400)
#supplementary alignment (0x800)


```